packages/transport-tcp/src/listener.ts

```diff
@@ -145,13 +153,22 @@ export class TCPListener extends EventEmitter<ListenerEvents> implements Listene
       })
       .on('close', () => {
         this.metrics?.status.update({
-          [this.addr]: SERVER_STATUS_DOWN
+          [this.addr]: this.status.code
         })
-        this.dispatchEvent(new CustomEvent('close'))
+
+        // If this event is emitted, the transport manager will remove the listener from it's cache
+        // in the meanwhile if the connections are dropped then listener will start listening again
+        // and the transport manager will not be able to close the server
+        if (this.status.code !== TCPListenerStatusCode.PAUSED) {
+          this.dispatchEvent(new CustomEvent('close'))
+        }
       })
   }
 
   private onSocket (socket: net.Socket): void {
+    if (this.status.code === TCPListenerStatusCode.DOWN) {
+      throw new Error('Server is is not listening yet')
```

Comment:
You can use the [CodeError](https://github.com/libp2p/js-libp2p/blob/1af6add837bff88e923faa4ffe4ad1af8e8783bd/packages/interface/src/errors.ts#L21) class